Added .net core 3.0 support

vincentparrett opened the pull request:
Added .net core 3.0 support - types were moved around and required referencing newer versions of the packages. I used the min version that worked (2.1) as netcore 3.0 is still in preview and didn't want to reference the preview packages (which change often).

vincentparrett commented:
Resolves issue #89 

Looks like you need to change your appveyor build config to build against netcore 3.0?

vincentparrett commented:
There are some issues with asp.net core 3.0 involving routing. At the moment, SmidgeStartup calls app.UseMvc to configure the routes, however in .netcore 3.0 you cannot do that if Endpoint routing is enabled (which will be the default). The problem is there does not appear to be a way (in UseSmidge) to tell whether the dev has endpointrouting enabled. I think the best bet is to only support use with endpoint routing, as it will be the default. For more info see here 
[https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/migration/22-to-30.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/migration/22-to-30.md)

Shazwazza commented:
Thanks for all your effort! I'll need to dive into this to understand it all but i will get there :) Again, many thanks for your input.

vincentparrett commented:
I have hit a roadblock on this. I cannot for the life of me get the routing to work. The taghelpers are working fine, but the controller never gets called. I'm not sure what the problem is, there are no errors, just get 404 on the bundle links. I have a feeling it may be a bung in preview 5, however trying the daily build for preview 6 just results in the debugger crashing. I'm going to put my .net core migration on hold until 3.0 stabilizes - just too much is fluid or broken right now.

vincentparrett commented:
Just pushed some changes that make it work with .net core 3.0 Preview 5.

Summary of changes

- Use Sdk="Microsoft.NET.Sdk.Web" - needed for asp.net core 3.0 
- Changed TargetFramework to TargetFrameworks and added netcoreapp3.0 target
- Added NETCORE3_0 define for netcoreapp3.0 targetframework
- Added useEndpointRouting to UseSmidge for netcoreapp3.0
- Added regControllersAsService to AddSmidge for for netcoreapp3.0

The method sig changes were needed for netcoreapp3.0 as there is no way to detect if EndpointRouting is turned off (it's on by default) - there is also no way to detect if AddControllersAsServices has been called before (if it has, we must too).

vincentparrett commented:
@Shazwazza this PR is ready for review, there are interface changes but only for .netcoreapp3.0 but since .net core 3.0 is still in preview you probably want to publish a preview version for now.

Shazwazza commented:
This is awesome! I will get to testing shortly

Shazwazza commented:
Testing and reviewing this now, sorry for the delay! great stuff :)

Shazwazza commented:
Have merged this in, there's a few things i'll fix up but it builds and runs, thanks a ton!!

Shazwazza commented:
Hi @vincentparrett ,

From all my tests this all seems to work great, the created package targets both `netstandard2.0` and `netcoreapp3.0` 

I've been out of the loop too long with net core 3 and what class libraries should be using as a target moniker in the nuget packages but I assume netcoreapp3.0 is correct? When i create a new class lib in VS with netcore3 this is the default target so I assume that's prob ok, any thoughts on that?

Is there any reason we should change the dependencies when targeting netcoreapp3.0 to be more like the netstandard2.0 build where it only references the packages it needs? or should it continue just reference the whole Microsoft.AspNetCore.App + Microsoft.NETCore.App ?

Apart from that it's basically ready to go. For this build since there's no new features/changes I'm going to release this under 3.1.0. A 4.0.0 is still being worked on that has a lot more fundamental changes. Do you think it's worth releasing a beta?

vincentparrett commented:
Microsoft changed how asp.net is referenced in netcore3.0, it's part of the sdk so you no longer need to reference all the packages. In this case we needed to change the project file to use the correct  sdk and .net core does the rest  when it comes to references. 

We've been using this version of smidge since I pushed this PR without any major issues, haven't checked with RC1 yet but will today. I would say it's fine to skip the beta, for .net framework and netcore 2.2 or earlier, there are no changes (use netstandard), for netcore 3.0 it requires a minor code change but that will be the least of your migration problems!

Shazwazza commented:
Aha, i see, this is a nice post about it https://andrewlock-net.cdn.ampproject.org/v/s/andrewlock.net/exploring-the-new-project-file-program-and-the-generic-host-in-asp-net-core-3/amp/?amp_js_v=a2&amp_gsa=1&usqp=mq331AQEKAFwAQ%3D%3D#aoh=15688459671871&referrer=https%3A%2F%2Fwww.google.com&amp_tf=From%20%251%24s&ampshare=https%3A%2F%2Fandrewlock.net%2Fexploring-the-new-project-file-program-and-the-generic-host-in-asp-net-core-3%2F

vincentparrett commented:
Actually now that I think about it, you should release this as a beta since .net core 3.0 is not released yet, still release candidate.

Shazwazza commented on src/Smidge/SmidgeStartup.cs:
```diff
@@ -85,27 +101,68 @@ public static class SmidgeStartup
 
+#if NETCORE3_0
+            //tell .net core 3.0 where to find our controller
+            var assembly = typeof(SmidgeStartup).Assembly;
```
I'm unsure about this block of code and the `regControllersAsService` parameter. It seems like just manually registering the controller would suffice? `services.TryAddTransient(typeof(SmidgeController));`

vincentparrett replied:
I found that if AddControllersAsServices had been called elsewhere, then you had to call it again otherwise it would not find the controllers. There is no way to detect if that had been done already (outside the call to AddSmidge). Just adding the controllers as transient services didn't work for me.. but that was back in preview5 so it's possible it will work now, worth testing.

Shazwazza replied:
Ok thanks, i'll investigate some more. Have been diving into the source code for some of this and starting to make some sense out of things. Just to be sure though, the reason calls would be made to AddControllersAsServices is if you opt in to use `ServiceBasedControllerActivator` instead of the default one?

Shazwazza replied:
Have had a look through all the source code for this stuff and doing what is here seems fine, though IMO we should remove the `regControllersAsService` and just ensure that people call this after they use `AddSmidge`. 

I was hoping it would be more simple to add our own 'parts' to the app and IMO it should be since what happens under the covers when we do this will work but it's also allocating a bunch more object instances and it's also re-checking a huge amount of stuff exists within the container. 

Do `AddController` does [this](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs#L87) and calls [AddControllersCore](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc/src/MvcServiceCollectionExtensions.cs#L141) which calls `AddMvcCore`, `AddApiExplorer`, `AddAuthorization`, `AddCors`, `AddDataAnnotations`, `AddFormatterMappings` and each of those methods goes on to do a huge amount of checks. 

At this stage, all we need is an instance of `IMvcCoreBuilder` and there's no simple way to get that without re-writing this method: [`GetApplicationPartManager`](https://github.com/aspnet/AspNetCore/blob/master/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L74) which is a bit ugly. 

So the code we can use that will have a minimal impact is this:

```cs
var builder = services.AddMvcCore();
builder.ConfigureApplicationPartManager(manager =>
{
    var assembly = typeof(SmidgeStartup).Assembly;
    var partFactory = ApplicationPartFactory.GetApplicationPartFactory(assembly);
    foreach (var applicationPart in partFactory.GetApplicationParts(assembly))
    {
        manager.ApplicationParts.Add(applicationPart);
    }
});

(which is similar code used by the [`AddApplicationPart`](https://github.com/aspnet/AspNetCore/blob/c7cb8467bfce721e2d66ef3862cd8c7c1fdbb421/src/Mvc/Mvc.Core/src/DependencyInjection/MvcCoreMvcBuilderExtensions.cs#L100) method)

With this change, it only ensure the minimum requirements to run smidge without enabling everything that it doesn't necessarily need. 
```

vincentparrett replied:
Looks good, I think. 

Does this mean that the call to addsmidge is now independent of position?
In my application, I'm calling :
```
services.AddSmidge()
  .AddControllersWithViews()
  .AddControllersAsServices() //for dependency injection as I'm using autofac
````
I remember there being issues with the smidge controllers not being found because AddControllersAsServices was called after AddSmidge, does this situation still work with the changes you made? I'm not able to debug this at the moment so can't tell for sure what the exact problem was... way to many balls in the air at the moment!

vincentparrett commented on src/Smidge/Models/BundleRequestModel.cs:
```diff
@@ -32,13 +32,13 @@ public BundleRequestModel(IUrlManager urlManager, IActionContextAccessor accesso
 
         public override ICacheBuster CacheBuster { get; }
-        public Bundle Bundle { get; }        
+        public Bundle FileBundle { get; }        
```
I had trouble compiling without this change. Perhaps that issue is resolved now, can't remember the exact detail now but it was a confilct with a type elsewhere I think.

Shazwazza replied:
Ok cool thanks i'll see what happens.

vincentparrett replied:
Looks like that change can be reverted as it compiles now with RC1
